src/node/internal/internal_zlib_base.ts

```diff
@@ -0,0 +1,558 @@
+                   handle.availInBefore, // in_len
+                   self._outBuffer, // out
+                   self._outOffset, // out_off
+                   self._chunkSize); // out_len
+        self._read(n);
+      };
+    }
+    return;
+  }
+
+  if (availInAfter > 0) {
+    // If we have more input that should be written, but we also have output
+    // space available, that means that the compression library was not
+    // interested in receiving more data, and in particular that the input
+    // stream has ended early.
+    // This applies to streams where we don't check data past the end of
+    // what was consumed; that is, everything except Gunzip/Unzip.
+    self.push(null);
+  }
+
+  // Finished with the chunk.
+  this.buffer = null;
+  this.cb();
+}
+
+// If a flush is scheduled while another flush is still pending, a way to figure
+// out which one is the "stronger" flush is needed.
+// This is currently only used to figure out which flush flag to use for the
+// last chunk.
+// Roughly, the following holds:
+// Z_NO_FLUSH (< Z_TREES) < Z_BLOCK < Z_PARTIAL_FLUSH <
+//     Z_SYNC_FLUSH < Z_FULL_FLUSH < Z_FINISH
+const flushiness: number[] = [];
+const kFlushFlagList = [CONST_Z_NO_FLUSH, CONST_Z_BLOCK, CONST_Z_PARTIAL_FLUSH,
+                        CONST_Z_SYNC_FLUSH, CONST_Z_FULL_FLUSH, CONST_Z_FINISH];
+for (let i = 0; i < kFlushFlagList.length; i++) {
+  flushiness[kFlushFlagList[i] as number] = i;
+}
+
+type BufferWithFlushFlag = Buffer & { [kFlushFlag]: number};
+
+// Set up a list of 'special' buffers that can be written using .write()
+// from the .flush() code as a way of introducing flushing operations into the
+// write sequence.
+const kFlushBuffers: BufferWithFlushFlag[] = [];
+{
+  const dummyArrayBuffer = new ArrayBuffer(0);
+  for (const flushFlag of kFlushFlagList) {
+    const buf = Buffer.from(dummyArrayBuffer) as BufferWithFlushFlag;
+    buf[kFlushFlag] = flushFlag;
+    kFlushBuffers[flushFlag] = buf;
+  }
+}
+
+function zlibOnError(this: zlibUtil.ZlibStream, errno: number, code: string, message: string): void {
+  const self = this[owner_symbol];
+  const error = new NodeError(code, message);
+  // @ts-expect-error Err number is expected.
+  error.errno = errno;
+  self.destroy(error);
+  self[kError] = error;
+}
+
+function processChunkSync(self: Zlib, chunk: Buffer, flushFlag: number): Buffer | Uint8Array {
+  let availInBefore = chunk.byteLength;
+  let availOutBefore = self._chunkSize - self._outOffset;
+  let inOff = 0;
+  let availOutAfter;
+  let availInAfter;
+
+  const buffers: (Buffer | Uint8Array)[] = [];
+  let nread = 0;
+  let inputRead = 0;
+  const state = self._writeState;
+  const handle = self._handle;
+  let buffer = self._outBuffer;
+  let offset = self._outOffset;
+  const chunkSize = self._chunkSize;
+
+  let error;
+  self.on('error', function onError(er) {
+    error = er;
+  });
+
+  while (true) {
+    // TODO: This was `writeSync` before, but it's not anymore.
```

Comment:
Is this our todo or an existing todo from the node.js source?

Reply:
This is our TODO. I implemented the `sync` path, not didn't implement the async path on C++.

Reply:
Ok, to differentiate, make the todo like... This will help since the `(soon)` is our convention vs. node.js convention of `TODO(username): `

```suggestion
    // TODO(soon): This was `writeSync` before, but it's not anymore.
```